paperkey.1
.\" paperkey manpage copyright (C) 2007, 2012, 2013 Peter Palfrader <peter@palfrader.org>
.\" Examples have been taken from David Shaw's README.
.\"
.\" This document is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This document is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
.\" MA 02110-1301 USA
.\"
.\"
.\" paperkey, the software, is written and
.\" copyright 2007, 2008 David Shaw <dshaw@jabberwocky.com>
.TH PAPERKEY 1 "June 2012" "PAPERKEY"
.SH NAME
paperkey \- extract secret information out of OpenPGP secret keys
.SH SYNOPSIS
.B paperkey\fR
[\fB\-\-secret\-key=\fR\fIFILE\fR]
[\fB\-\-output=\fR\fIFILE\fR]
[\fB\-\-output\-type=\fR\fBbase16\fR|\fBraw\fR]
[\fB\-\-output\-width=\fR\fIWIDTH\fR]
.LP
.B paperkey\fR
\fB\-\-pubring=\fR\fIFILE\fR
[\fB\-\-secrets=\fR\fIFILE\fR]
[\fB\-\-input\-type=\fR\fBauto\fR|\fBbase16\fR|\fBraw\fR]
[\fB\-\-output=\fR\fIFILE\fR]
[\fB\-\-ignore\-crc\-error\fR]
[\fB\-\-comment=\fR\fISTRING\fR]
[\fB\-\-file\-format\fR]
.LP
.B paperkey\fR \fB\-\-version\fR
.SH MOTIVATION
As with all data, secret keys should be backed up.  In fact, secret
keys should be backed up even better than other data, because they are
impossible to recreate should they ever be lost.  All files encrypted
to lost keys are forever (or at least for a long time) undecipherable.
In addition to keeping backups of secret key information on digital
media such as USB-sticks or CDs it is reasonable to keep an
if-all-else-fails copy on plain old paper, for use should your digital
media ever become unreadable for whatever reason.  Stored properly,
paper is able to keep information for several decades or longer.
.PP
With GnuPG, PGP, or other OpenPGP implementations the secret key
usually contains a lot more than just the secret numbers that are
important.  They also hold all the public values of key pairs, user
ids, expiration times and more.  In order to minimize the information
that has to be entered manually or with the help of OCR, QR code or
similar software, \fBpaperkey\fR extracts just the secret information
out of OpenPGP secret keys.  For recovering a secret key it is assumed
that the public key is still available, for instance from public
Internet keyservers.
.SH DESCRIPTION
\fBpaperkey\fR has two modes of operation:
.PP
The first mode creates "paperkeys" by extracting just the secret
information from a secret key, formatting the data in a way suitable
for printing or in a raw mode for further processing.
.PP
The other mode rebuilds secret keys from such a paperkey and a copy of
the public key, also verifying the checksums embedded in the paperkey.
This mode is selected when the \fB\-\-pubring\fR option is used, which
is required in that case.  If a passphrase was set on the original
secret key, the same passphrase is set on the rebuilt key.
.PP
Input is read from standard\-in except when the \fB\-\-secret\-key\fR or
\fB\-\-secrets\fR option is used; output is printed to standard\-out,
unless changed with the \fB\-\-output\fR option.
.SH SECURITY CONSIDERATIONS
Please note that \fBpaperkey\fR does not change the protection and
encryption status of and security requirements for storing your secret
key. If the secret key was protected by a passphrase so is the
paperkey.  If the secret key was unprotected the paperkey will not be
protected either.
.SH OPTIONS
\fB\-\-help\fR, \fB\-h\fR
Display a short help message and exit successfully.
.LP
.TP
\fB\-\-version\fR, \fB\-V\fR
Print version information and copyright information and exit successfully.
.LP
.TP
\fB\-\-verbose\fR, \fB\-v\fR
Print status and progress information to standard\-error while processing
the input.  Repeat for even more output.
.LP
.TP
\fB\-\-output=\fR\fIFILE\fR, \fB\-o\fR
Redirect output to the file given instead of printing to standard\-output.
.LP
.TP
\fB\-\-comment=\fR\fISTRING\fR
Include the specified comment in the base16 output.
.LP
.TP
\fB\-\-file\-format\fR
Paperkey automatically includes the file format it uses as comments at
the top of the base16 output.  This command simply prints out the file
format and exits successfully.
.SH OPTIONS FOR EXTRACTING SECRET INFORMATION
.TP
\fB\-\-output\-type=base16\fR, \fB\-\-output\-type=raw\fR
Select the output type.  The base16 style encodes the information in
the style of a classic hex-dump, including line numbers and per-line
CRC checksums to facilitate localizing errors in the input file during
the recovery phase.  The raw, or binary, mode is just a raw dump of
the secret information, intended for feeding to barcode generators or
the like.
.LP
.TP
\fB\-\-output\-width=\fR\fIWIDTH\fR
Choose line width in the base16 output mode.  The default is 78 characters.
.LP
.TP
\fB\-\-secret\-key=\fR\fIFILE\fR
File to read the secret key from.  If this option is not given \fBpaperkey\fR
reads from standard\-input.
.SH OPTIONS FOR RE-CREATING PRIVATE KEYS
.TP
\fB\-\-input\-type=auto\fR, \fB\-\-input\-type=base16\fR, \fB\-\-input\-type=raw\fR
Specify that the given input is either in base16 format, as produced
by \fBpaperkey\fR, or in raw format.  The default, auto, tries to
automatically detect the format in use.
.LP
.TP
\fB\-\-pubring=\fR\fIFILE\fR
File to read public key information from.  It is assumed that the user can
get the public key from sources like public Internet keyservers.
.LP
.TP
\fB\-\-secrets=\fR\fIFILE\fR
File to read the extracted secrets, the paperkey, from.  If this is not given
then the information is read from standard\-input.
.LP
.TP
\fB\-\-ignore\-crc\-error\fR
Do not reject corrupt input and continue despite any CRC errors.
.SH EXAMPLES
Take the secret key in key.gpg and generate a text file
to\-be\-printed.txt that contains the secret data:
.PP
.B $ paperkey \-\-secret\-key my\-secret\-key.gpg \-\-output to\-be\-printed.txt
.PP
Take the secret key data in my\-key\-text\-file.txt and combine it with
my\-public\-key.gpg to reconstruct my\-secret\-key.gpg:
.PP
.B $ paperkey \-\-pubring my\-public\-key.gpg \-\-secrets my\-key\-text\-file.txt \-\-output my\-secret\-key.gpg
.PP
If \-\-output is not specified, the output goes to stdout.  If \-\-secret\-key is
not specified, the data is read from stdin so you can do things like:
.PP
.B $ gpg \-\-export\-secret\-key my\-key | paperkey | lpr
.SH SEE ALSO
.BR gpg (1),
.BR http://www.jabberwocky.com/software/paperkey/
.SH AUTHORS
\fBpaperkey\fR is written by David Shaw <dshaw@jabberwocky.com>.


NEWS
Noteworthy changes in version 1.4 (unreleased)
----------------------------------------------


Noteworthy changes in version 1.3 (2013-01-03)
----------------------------------------------

    * Added support for OpenPGP elliptic curve keys (ECDH and ECDSA)
      as specified in RFC 6637.

    * Use minimal OpenPGP packet encoding.  This does not affect
      functionality, but makes it more likely that a given key will
      match byte-for-byte the same key after running through paperkey.

    * Minor bug fixes.

    * Gnulib refresh.


Noteworthy changes in version 1.2 (2009-10-14)
----------------------------------------------

    * Minor documentation changes.

    * Gnulib refresh.

    * Generate output files with more restrictive permissions (go-rwx).


Noteworthy changes in version 1.1 (2009-02-10)
----------------------------------------------

    * Minor bug fix for Windows platforms.

    * Make the instructions on how to restore a key without paperkey a
      bit clearer.


Noteworthy changes in version 1.0 (2009-01-22)
----------------------------------------------

    * Minor documentation and build changes.

    * Gnulib refresh.


Noteworthy changes in version 0.8 (2008-02-01)
----------------------------------------------

    * The file format is now included as part of the base16 output, as
      there is no guarantee that this program will be on-hand when a
      reconstruction is necessary.  The format can also be displayed
      via the --file-format command.  Suggested by Brendan Kidwell.

    * Some bug fixes (actually to gnulib, but relevant here as well)
      to the SHA-1 code on platforms that require aligned access.
      Thanks to Peter Palfrader.

    * New --comment option to add comments to the base16 output.


Noteworthy changes in version 0.7 (2007-10-03)
----------------------------------------------

    * Man page contributed by Peter Palfrader.

    * Secret data can now be passed in on standard input if desired.


Noteworthy changes in version 0.6 (2007-09-22)
----------------------------------------------

    * Bug fix for certain key sizes thanks to Peter Palfrader.

    * Add versioning to the output format.

    * Add an overall CRC to the secrets output to catch certain types
      of corruption.

    * Properly use a single public key out of a whole public keyring
      when restoring.


Noteworthy changes in version 0.5 (2007-09-21)
----------------------------------------------

    * Initial release.


paperkey.c
/*
 * Copyright (C) 2007, 2008, 2009, 2012 David Shaw <dshaw@jabberwocky.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <config.h>
#include <stdio.h>
#include <getopt.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#ifdef _WIN32
#include <io.h>
#include <fcntl.h>
#endif
#include "output.h"
#include "extract.h"
#include "restore.h"

int verbose=0,ignore_crc_error=0;
unsigned int output_width=78;
char *comment=NULL;

enum options
  {
    OPT_HELP=256,
    OPT_VERSION,
    OPT_VERBOSE,
    OPT_OUTPUT,
    OPT_INPUT_TYPE,
    OPT_OUTPUT_TYPE,
    OPT_OUTPUT_WIDTH,
    OPT_SECRET_KEY,
    OPT_PUBRING,
    OPT_SECRETS,
    OPT_IGNORE_CRC_ERROR,
    OPT_FILE_FORMAT,
    OPT_COMMENT
  };

static struct option long_options[]=
  {
    {"help",no_argument,NULL,OPT_HELP},
    {"version",no_argument,NULL,OPT_VERSION},
    {"verbose",no_argument,NULL,OPT_VERBOSE},
    {"output",required_argument,NULL,OPT_OUTPUT},
    {"input-type",required_argument,NULL,OPT_INPUT_TYPE},
    {"output-type",required_argument,NULL,OPT_OUTPUT_TYPE},
    {"output-width",required_argument,NULL,OPT_OUTPUT_WIDTH},
    {"secret-key",required_argument,NULL,OPT_SECRET_KEY},
    {"pubring",required_argument,NULL,OPT_PUBRING},
    {"secrets",required_argument,NULL,OPT_SECRETS},
    {"ignore-crc-error",no_argument,NULL,OPT_IGNORE_CRC_ERROR},
    {"file-format",no_argument,NULL,OPT_FILE_FORMAT},
    {"comment",required_argument,NULL,OPT_COMMENT},
    {NULL,0,NULL,0}
  };

static void
usage(void)
{
  printf("Usage: paperkey [OPTIONS]\n");
  printf("  --help (-h)\n");
  printf("  --version (-V)\n");
  printf("  --verbose (-v)  be more verbose\n");
  printf("  --output (-o)   write output to this file\n");
  printf("  --input-type    auto, base16 or raw (binary)\n");
  printf("  --output-type   base16 or raw (binary)\n");
  printf("  --output-width  maximum width of base16 output\n");
  printf("  --secret-key"
	  "    extract secret data from this secret key\n");
  printf("  --pubring"
	  "       public keyring to find non-secret data\n");
  printf("  --secrets       file containing secret"
	  " data to join with the public key\n");
  printf("  --ignore-crc-error  don't reject corrupted input\n");
  printf("  --file-format   show the paperkey file format\n");
  printf("  --comment       add a comment to the base16 output\n");
}

int
main(int argc,char *argv[])
{
  int arg,err;
  FILE *secret_key,*secrets,*pubring=NULL;
  const char *outname=NULL;
  enum data_type output_type=BASE16;
  enum data_type input_type=AUTO;

#ifdef _WIN32
  if(_setmode(_fileno(stdin),_O_BINARY)==-1)
    {
      fprintf(stderr,"Unable to set stdin mode to binary: %s\n",
	      strerror(errno));
      exit(1);
    }
#endif

  secret_key=secrets=stdin;

  /* Force the umask to go-rwx, as we are going to be creating files
     that contain secret key material. */
  umask(077);

  while((arg=getopt_long(argc,argv,"hVvo:",long_options,NULL))!=-1)
    switch(arg)
      {
      case OPT_HELP:
      case 'h':
      default:
        usage();
        exit(0);

      case OPT_VERSION:
      case 'V':
	printf("%s (%s)\n",PACKAGE_STRING,HOST_OS);
	printf("%s\n",COPYRIGHT_STRING);
	printf("This is free software.  You may redistribute copies of it"
	       " under the terms of\n");
	printf("the GNU General Public License"
	       " <http://www.gnu.org/licenses/gpl.html>.\n");
	printf("There is NO WARRANTY, to the extent permitted by law.\n\n");
	exit(0);

      case OPT_VERBOSE:
      case 'v':
	verbose++;
	break;

      case OPT_OUTPUT:
      case 'o':
	outname=optarg;
	break;

      case OPT_INPUT_TYPE:
	if(strcmp(optarg,"auto")==0)
	  input_type=AUTO;
	else if(strcmp(optarg,"base16")==0)
	  input_type=BASE16;
	else if(strcmp(optarg,"raw")==0)
	  input_type=RAW;
	else
	  {
	    fprintf(stderr,"Unknown input type \"%s\"\n",optarg);
	    exit(1);
	  }
	break;

      case OPT_OUTPUT_TYPE:
	if(strcmp(optarg,"base16")==0)
	  output_type=BASE16;
	else if(strcmp(optarg,"raw")==0)
	  output_type=RAW;
	else
	  {
	    fprintf(stderr,"Unknown output type \"%s\"\n",optarg);
	    exit(1);
	  }
	break;

      case OPT_OUTPUT_WIDTH:
	output_width=atoi(optarg);
	break;

      case OPT_SECRET_KEY:
	secret_key=fopen(optarg,"rb");
	if(!secret_key)
	  {
	    fprintf(stderr,"Unable to open %s: %s\n",optarg,strerror(errno));
	    exit(1);
	  }
	break;

      case OPT_PUBRING:
	pubring=fopen(optarg,"rb");
	if(!pubring)
	  {
	    fprintf(stderr,"Unable to open pubring %s: %s\n",
		    optarg,strerror(errno));
	    exit(1);
	  }
	break;

      case OPT_SECRETS:
	secrets=fopen(optarg,"rb");
	if(!secrets)
	  {
	    fprintf(stderr,"Unable to open %s: %s\n",optarg,strerror(errno));
	    exit(1);
	  }
	break;

      case OPT_IGNORE_CRC_ERROR:
	ignore_crc_error=1;
	break;

      case OPT_FILE_FORMAT:
	output_file_format(stdout,"");
	exit(0);

      case OPT_COMMENT:
	comment=optarg;
	break;
      }

  if(pubring)
    err=restore(pubring,secrets,input_type,outname);
  else
    err=extract(secret_key,outname,output_type);

  return err;
}
